packages/truffle-decoder/lib/decode/index.ts

```diff
@@ -9,45 +9,52 @@ import { decodeLiteral } from "./stack";
 import decodeCalldata from "./calldata";
 import decodeConstant from "./constant";
 import decodeSpecial from "./special";
-import { AstDefinition } from "truffle-decode-utils";
+import { AstDefinition, Types, Values } from "truffle-decode-utils";
 import * as Pointer from "../types/pointer";
 import { EvmInfo } from "../types/evm";
-import { DecoderRequest } from "../types/request";
+import { DecoderRequest, GeneratorJunk } from "../types/request";
 
-export default function* decode(definition: AstDefinition, pointer: Pointer.DataPointer, info: EvmInfo): IterableIterator<any | DecoderRequest> {
-  debug("Decoding %s", definition.name);
+export default function* decode(definition: AstDefinition, pointer: Pointer.DataPointer, info: EvmInfo): IterableIterator<Values.Result | DecoderRequest | GeneratorJunk> {
```

Comment:
You know, I kind of almost want to say we should force `definitionToType` to be outside `decode`. It looks like we really don't use definitions that much here anymore, may be best to just do away with them for `lib/decode`? Unless I'm missing something.

Reply:
Hm, so, let me tell you how I have it right now on the `wire-decoder` branch, because I've reworked this a tiny bit there, in a way that I think you might prefer.  There, the function that's called `decodePointer` here is just called `decode`.  The function that's here called `decode`, that performs the `definitionToType` and then dispatches to `decode`/`decodePointer`, is there called `decodeVariable`.  Only it doesn't live in `lib/decode/index.ts`; rather it lives in `lib/interface/decoding.ts`, along with `decodeCalldata` and `decodeEvent`, which have their own ways of calling `decode` (aka `decodePointer`).

So, it sounds to me like I've already created a structure more in line with what you want, only I did it on `wire-decoder` rather than `type-and-value`.  So, we could wait for `wire-decoder`, or I could backport that change if you think it's worth it.

Reply:
Oh, nice! Yeah I like that a lot more. Definitely we can wait, then. It'll be the same breaking release, anyway.